File: native-wrappers/android/datadogwrapper/src/main/kotlin/com/datadog/wrapper/DdTelemetry.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.wrapper

import com.datadog.android.Datadog
import com.datadog.android._InternalProxy
import com.datadog.android.event.EventMapper
import com.datadog.android.rum.RumConfiguration
import com.datadog.android.rum._RumInternalProxy
import com.datadog.android.telemetry.model.TelemetryConfigurationEvent
import java.util.concurrent.ConcurrentHashMap

/**
 * Bridge for SDK-internal telemetry. Forwards wrapper-level errors, debug
 * messages, and configuration metadata into the Android SDK's telemetry pipeline,
 * so they end up alongside native telemetry in the Datadog backend.
 *
 * Configuration fields are accumulated as the C# layer reports them at SDK
 * init / RUM enable / Trace enable. The accumulated fields are applied to the
 * `TelemetryConfigurationEvent` via a mapper installed on `RumConfiguration.Builder`
 * during `DdRum.enableRum`. This mirrors how dd-sdk-flutter and dd-sdk-reactnative
 * augment the native config event with cross-platform metadata.
 *
 * All entry points are no-ops if Datadog is not initialized.
 */
class DdTelemetry {
    companion object {

        // Dependencies - injectable for testing
        private var telemetryProvider: TelemetryProvider = RealTelemetryProvider()

        // Accumulator for cross-platform configuration fields.
        // Keys here match the keys used in the C# `InternalTelemetry.ReportConfiguration`
        // call site. The mapper below reads from this map at event-emission time.
        private val pendingConfiguration = ConcurrentHashMap<String, Any>()

        @JvmStatic
        fun setTelemetryProvider(provider: TelemetryProvider) {
            telemetryProvider = provider
        }

        // Reset to production dependencies (for test cleanup)
        @JvmStatic
        fun resetDependencies() {
            telemetryProvider = RealTelemetryProvider()
            pendingConfiguration.clear()
        }

        // -- Error / Debug --

        @JvmStatic
        @JvmOverloads
        fun error(message: String, kind: String? = null, stack: String? = null) {
            val proxy = telemetryProvider.telemetryProxy ?: return
            proxy.error(message, stack, kind)
        }

        @JvmStatic
        fun errorWithThrowable(message: String, throwable: Throwable?) {
            val proxy = telemetryProvider.telemetryProxy ?: return
            proxy.error(message, throwable)
        }

        @JvmStatic
        fun debug(message: String) {
            val proxy = telemetryProvider.telemetryProxy ?: return
            proxy.debug(message)
        }

        // -- Configuration telemetry --

        /**
         * Accumulates cross-platform configuration fields. Applied to emitted
         * `TelemetryConfigurationEvent` events by the mapper installed in
         * [installConfigurationMapper]. Fields not in the recognized key set
         * are stored but ignored at emission time.
         */
        @JvmStatic
        fun reportConfiguration(fields: Map<String, Any?>) {
            for ((key, value) in fields) {
                if (value != null) {
                    pendingConfiguration[key] = value
                }
            }
        }

        /**
         * Installs the telemetry-configuration mapper on the given builder.
         * Must be called from `DdRum.enableRum` before `.build()`. Without this
         * hook, accumulated configuration fields aren't applied to events.
         */
        @JvmStatic
        fun installConfigurationMapper(builder: RumConfiguration.Builder): RumConfiguration.Builder {
            return _RumInternalProxy.setTelemetryConfigurationEventMapper(
                builder,
                object : EventMapper<TelemetryConfigurationEvent> {
                    override fun map(event: TelemetryConfigurationEvent): TelemetryConfigurationEvent {
                        applyPendingConfiguration(event.telemetry.configuration)
                        return event
                    }
                }
            )
        }

        private fun applyPendingConfiguration(config: TelemetryConfigurationEvent.Configuration) {
            // Strings
            (pendingConfiguration["mauiVersion"] as? String)?.let { config.mauiVersion = it }
            (pendingConfiguration["initializationType"] as? String)?.let { config.initializationType = it }
            (pendingConfiguration["tracerAPI"] as? String)?.let { config.tracerApi = it }

            // Booleans
            (pendingConfiguration["trackErrors"] as? Boolean)?.let { config.trackErrors = it }
            (pendingConfiguration["trackNativeErrors"] as? Boolean)?.let { config.trackNativeErrors = it }
            (pendingConfiguration["trackUserInteractions"] as? Boolean)?.let { config.trackUserInteractions = it }
            (pendingConfiguration["trackResources"] as? Boolean)?.let { config.trackResources = it }
            (pendingConfiguration["trackBackgroundEvents"] as? Boolean)?.let { config.trackBackgroundEvents = it }
            (pendingConfiguration["trackFrustrations"] as? Boolean)?.let { config.trackFrustrations = it }
            (pendingConfiguration["trackNativeViews"] as? Boolean)?.let { config.trackNativeViews = it }
            (pendingConfiguration["trackViewsManually"] as? Boolean)?.let { config.trackViewsManually = it }
            (pendingConfiguration["trackLongTask"] as? Boolean)?.let { config.trackLongTask = it }
            (pendingConfiguration["trackNativeLongTasks"] as? Boolean)?.let { config.trackNativeLongTasks = it }
            (pendingConfiguration["useFirstPartyHosts"] as? Boolean)?.let { config.useFirstPartyHosts = it }
            (pendingConfiguration["useProxy"] as? Boolean)?.let { config.useProxy = it }
        }
    }

    /**
     * Provider of the `_TelemetryProxy` handle. Allows test injection.
     */
    interface TelemetryProvider {
        val telemetryProxy: _InternalProxy._TelemetryProxy?
    }

    private class RealTelemetryProvider : TelemetryProvider {
        override val telemetryProxy: _InternalProxy._TelemetryProxy?
            get() = if (Datadog.isInitialized()) {
                Datadog._internalProxy()._telemetry
            } else {
                null
            }
    }
}

File: native-wrappers/ios/DatadogWrapper/Sources/DatadogWrapper/DdTelemetry.swift
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

import Foundation
import DatadogCore
import DatadogInternal

/// Bridge for SDK-internal telemetry. Forwards wrapper-level errors, debug
/// messages, and configuration metadata into the iOS SDK's telemetry pipeline,
/// so they end up alongside native telemetry in the Datadog backend.
///
/// All entry points are no-ops if Datadog is not initialized.
@objc(DdTelemetry)
public class DdTelemetry: NSObject {

    // Dependencies - injectable for testing
    private static var telemetryProvider: TelemetryProviderProtocol = RealTelemetryProvider()

    static func setTelemetryProvider(_ provider: TelemetryProviderProtocol) {
        telemetryProvider = provider
    }

    // Reset to production dependencies (for test cleanup)
    static func resetDependencies() {
        telemetryProvider = RealTelemetryProvider()
    }

    // MARK: - Error / Debug

    @objc public static func error(id: String, message: String, kind: String?, stack: String?) {
        guard let telemetry = telemetryProvider.telemetry else { return }
        telemetry.error(
            id: id,
            message: message,
            kind: kind ?? "MauiSdkError",
            stack: stack ?? ""
        )
    }

    @objc public static func debug(id: String, message: String) {
        guard let telemetry = telemetryProvider.telemetry else { return }
        telemetry.debug(id: id, message: message)
    }

    // MARK: - Configuration telemetry

    /// Reports configuration telemetry. Fields are read out of the dictionary using
    /// keys that match the corresponding `Telemetry.configuration(...)` parameters.
    /// Unknown/missing keys are simply omitted.
    /// Argument order must match the SDK's `Telemetry.configuration(...)` declaration.
    @objc public static func reportConfiguration(_ fields: NSDictionary) {
        guard let telemetry = telemetryProvider.telemetry else { return }

        let dict = fields as? [String: Any] ?? [:]
        telemetry.configuration(
            appHangThreshold: (dict["appHangThreshold"] as? NSNumber)?.int64Value,
            initializationType: dict["initializationType"] as? String,
            sessionSampleRate: (dict["sessionSampleRate"] as? NSNumber)?.int64Value,
            telemetryConfigurationSampleRate: (dict["telemetryConfigurationSampleRate"] as? NSNumber)?.int64Value,
            telemetrySampleRate: (dict["telemetrySampleRate"] as? NSNumber)?.int64Value,
            tracerAPI: dict["tracerAPI"] as? String,
            traceSampleRate: (dict["traceSampleRate"] as? NSNumber)?.int64Value,
            trackBackgroundEvents: dict["trackBackgroundEvents"] as? Bool,
            trackErrors: dict["trackErrors"] as? Bool,
            trackFrustrations: dict["trackFrustrations"] as? Bool,
            trackLongTask: dict["trackLongTask"] as? Bool,
            trackNativeErrors: dict["trackNativeErrors"] as? Bool,
            trackNativeLongTasks: dict["trackNativeLongTasks"] as? Bool,
            trackNativeViews: dict["trackNativeViews"] as? Bool,
            trackResources: dict["trackResources"] as? Bool,
            trackUserInteractions: dict["trackUserInteractions"] as? Bool,
            trackViewsManually: dict["trackViewsManually"] as? Bool,
            mauiVersion: dict["mauiVersion"] as? String,
            useFirstPartyHosts: dict["useFirstPartyHosts"] as? Bool,
            useProxy: dict["useProxy"] as? Bool,
            useTracing: dict["useTracing"] as? Bool
        )
    }
}

/// Provider of the Telemetry handle. Allows test injection.
protocol TelemetryProviderProtocol {
    var telemetry: Telemetry? { get }
}

class RealTelemetryProvider: TelemetryProviderProtocol {
    var telemetry: Telemetry? {
        // CoreRegistry.default.telemetry returns a NOP if uninitialized,
        // so we can safely return it; callers do not need to check.
        return CoreRegistry.default.telemetry
    }
}
